REST_API_COMPATIBILITY.md

```diff
@@ -0,0 +1,209 @@
+
+The version in `forRestApiVersion` is reference to when the declaration is valid. Assuming version 8 is the master branch and all changes start in the master branch then get back ported. The above text is what would be applicable for the v8 branch of code. The first line of code is essentially ignored except for when compatibility with version 7 is requested. When back-porting this change to the 7.x branch, the first line would be identical, and the second line would be omitted.
+
+The above strategy works well for single fields, but could get overly complex very fast for large multiple field changes. For more complex de-serialization changes there is also support to construct a `NamedXContentRegistry` with some "normal" entries as well some entries that are only applied when compatibility with the prior version is requested. The syntax is very similar where can express the desired version from the required `ParseField` when adding an entry to the `NamedXContentRegistry`.
+
+Additionally, there is a pseudo standard method `fromXContent` which also generally has access to the `XContentParser` for that request which can be used to conditionally change how to parse the input.
+
+```java
+private static final ParseField limitField = new ParseField("maximum", "limit").forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7));
+
+//call to fromXContent
+MyExample.fromXContent(XContentType.JSON.xContent().createParser(XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE).withRestApiVersion(request.getRestApiVersion()), " { \"limit\" : 99 }"));
+
+//contents of a fromXContent
+while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+    if (token == XContentParser.Token.FIELD_NAME) {
+        currentFieldName = parser.currentName();
+    } else if (token.isValue()) {
+        if (limitField.match(currentFieldName, LoggingDeprecationHandler.INSTANCE)) {
+            if (parser.getRestApiVersion().matches(RestApiVersion.onOrAfter(RestApiVersion.V_8))
+                && "maximum".equals(currentFieldName) == false) {
+                throw new IllegalArgumentException("invalid parameter [limit], use [maximum] instead");
+            } else {
+                value = parser.intValue();
+            }
+        }
+    }
+}
+```
+### URL
+
+Paths are declared via a Route. Routes are composed of the HTTP verb and the relative path. Optionally they can declare a deprecated verb/path combination and the REST API version in which they were deprecated.
+
+For example:
+
+```java
+   Route.builder(GET, "_mypath/{foo}/{bar}").deprecated(MY_DEPRECATION_MESSAGE, RestApiVersion.V_7).build(),
+```
+
```

Comment:
```suggestion
Note: you can safely backport this route declaration to 7.x branch.
```
not sure we should mention what should is safe to be backported (route declarations are, but compatible code probably isn't.

edit:.. We mention about this in developer's workflow, so probably not needed